### ISSUE TITLE
Returning functions in place

### ISSUE THREAD

**Issue by szuckerman** [ericmjl/pyjanitor#76]
After seeing issue #67, was curious as to what people think as to adding this capability to all functions?

Some of them, like `df.limit_column_characters()`, already return in place. I don't think it will be hard to extend to others.

**Comment by ericmjl** [comment]
:thinking: :thinking: :thinking: 

This one is a tough one for me to respond to.

On one hand, I can see why that might be desirable by some users who are used to the `inplace=True` kwarg.

On the other hand, I think adding it in as an option means an additional thing to check for when any new function is submitted in. Conceptually, it's "just one less thing" to maintain if we just leave that argument out.

With a [fluent API](https://en.wikipedia.org/wiki/Fluent_interface), I don't think in-place modifications are a thing. 

That said, I'm willing to be persuaded by a counter-example to my general slant against in-place modification of a dataframe, if one can provide an example where in-place modification accomplishes something substantial that a method chaining interface cannot accomplish.

**Comment by szuckerman** [comment]
I’m actually more inclined have the default be to return functions in place this is for a few reasons:

1. It doesn’t change the method chaning syntax
2. If people want to apply separate functions outside of the initial janitor section it’s merely ‘df.add_column()’ rather than ‘df=df.add_column’. I think the former is cleaner. 
3. Since Pandas is used for a variety of reasons, with EDA being a big one, it could be a big problem to change dataframes in place since one’s work could be inadvertently clobbered. My assumption is that the janitor functions should run once at the beginning and then later people will work with Pandas after the data is cleaned. If someone did clobber a dataframe, I would assume he would just “start over”

**Comment by ericmjl** [comment]
@szuckerman, let me see if I can clarify my understanding of the definition of in-place vs. method chaining.

Taking the example from the README, regular `pandas` syntax looks as follows:

```python
df = pd.DataFrame(...)  # create a pandas DataFrame somehow.
del df['column1']  # delete a column from the dataframe.
df = df.dropna(subset=['column2', 'column3'])  # drop rows that have empty values in column 2 and 3.
df = df.rename({'column2': 'unicorns', 'column3': 'dragons'})  # rename column2 and column3
df['newcolumn'] = ['iterable', 'of', 'items']  # add a new column.
```

First off, with in-place modifications at each line, we do:

```python
df = pd.DataFrame(...)
df.add_column()
```

**Comment by ericmjl** [comment]
> That being the case, some of the functions already work in place.

@szuckerman yes, that's a valid concern, especially for consistency purposes. 

I think the hack that we've implicitly used, but not explicitly stated, is to piggy back off pandas operations. If modification takes place in-place anyways (to prevent memory copying), then it happens in-place; if not, then we don't really care. For a pareto-distributed set of users, 80% will have datasets that fit in-memory, while 20% will have @zbarry's kind of needs, where dataframes are too large for a single computer's RAM.

At the moment, I don't think we're necessarily at a point where too many people will complain about the memory-usage architecture of `pyjanitor`. Perhaps when a few related issues come together, then it might be an issue? (I will defer to your experience, @szuckerman, as the more seasoned software engineer here, as to whether we should enforce consistency w.r.t. memory usage at this early point in the project.)

**Comment by szuckerman** [comment]
@zbarry makes a good point that "since you can't get the intermediate dataframes out for each step anyway" returning in place is more performant.

However, after doing some research [it doesn't look like returning objects in place makes too much of a difference](https://stackoverflow.com/questions/45570984/pandas-is-inplace-true-considered-harmful-or-not). I think the method-chaining syntax saves most people from the concern of making mistakes (i.e. by accidentally running a method that works in place by default), as given as one of the examples on stackoverflow.

As for future considerations of the package, I think that to really determine whether methods should default to being in place or not depends on "usability". Will users be confused when dataframes are mutated in place?  Later we can run tests, profile memory, etc. to determine if there's a difference between a method and its in place counterpart.

**Comment by zbarry** [comment]
Yeah, usability is probably core to the idea of pyjanitor even. That being said, I think having a (short) paragraph in the readme + in the notebook example I will (eventually) create as well as having a clear boilerplate message in all the function docs would convey this pretty well. If you make it such that all functions that are relevant w.r.t. this discussion operate on the same principle of in place, I think this gives a clear expection to users as to what approach pyjanitor takes for operations on `DataFrame`s (that being that they should expect mutation).

Because I'm procrastinating, I ran a simple speed test for adding a column to a `DataFrame` with `pandas`' `.assign` vs. using the `df[my_col] = data` explicitly (which is what `pyjanitor`s `.add_column` does):

![image](https://user-images.githubusercontent.com/4762711/49523810-7eaf9400-f878-11e8-8048-eded654520c0.png)

The slowdown with the data copy in this particular operation is a pretty consistent 1.5x. This could definetely add up when users chain multiple operations (I think it's possible that some ops could have even worse slowdown multipliers).

**Comment by zbarry** [comment]
Slowdown multiplier when adding to a `DataFrame` that has multiple columns already is pretty horrible:

![image](https://user-images.githubusercontent.com/4762711/49524922-a9025100-f87a-11e8-8641-be38112774ac.png)

**Comment by ericmjl** [comment]
@zbarry those are great tests. Yes, I think it's worth adding to the docstrings (but without user warnings popping up - those are too easily abused) that for the relevant functions, underneath the hood, we do change the original dataframe in place for performance reasons.

**Comment by zbarry** [comment]
> without user warnings popping up

Hmm... I was thinking of adding a pyqt5 dependency & popping up a message box warning you upon pyjanitor import of the mutation operations. Shame.

**Comment by zbarry** [comment]
Random idea: what about having colorful badges in the docs that you can see at a glance whether there is mutation / something else critical to know that a function does to your dataframe?

Like:
![image](https://user-images.githubusercontent.com/4762711/49690188-97c77780-fafa-11e8-94b1-3027863f2463.png)
kind of stuff but with appropriate labels.
